### PR TITLE
Tcota 431

### PR DESCRIPTION
**Pull request by isururanawaka**

### CODE REVIEW

**Review by machristie — CHANGES_REQUESTED** [review]
Hi @isururanawaka , thanks and I think this represents some good progress. However I don't think this is quite there. See my review comments.

Also, while this PR checks if the submissions limit has been exceeded when a submission is made, we also need a check to see if the user can create a submission at the beginning of the process. See StepsController: LAUNCH_ASSIGNMENT which is called to create a new submission record for the student when they first load the assignment. We want to let the student know before they attempt to do the assignment that they have maxed out their attempts.

File: src/main/java/edu/iu/terracotta/controller/app/StepsController.java [hunk 1]
```diff
@@ -81,75 +68,83 @@ public class StepsController {
...
+                    } else if (apijwtService.isInstructorOrHigher(securedInfo)) {
+                        for (String submissionIdString : submissionsId) {
                             Long submissionId = Long.parseLong(submissionIdString);
                             submissionService.finalizeAndGrade(submissionId, securedInfo);
                         }
+                    } else {
+                        return new ResponseEntity<>(TextConstants.NOT_ENOUGH_PERMISSIONS, HttpStatus.UNAUTHORIZED);
+                    }
                 } else {
                     return new ResponseEntity<>(TextConstants.NOT_ENOUGH_PERMISSIONS, HttpStatus.UNAUTHORIZED);
```
Comment (machristie) [inline src/main/java/edu/iu/terracotta/controller/app/StepsController.java:131]: Can we use the `MAX_SUBMISSION_ATTEMPTS_REACHED` message here too?

File: src/main/java/edu/iu/terracotta/service/app/impl/QuestionSubmissionServiceImpl.java [hunk 2]
```diff
@@ -304,4 +306,38 @@ public void validateQuestionSubmission(QuestionSubmissionDto questionSubmissionD
...
+                        canvasCourseId, platformDeployment);
+
+                Optional<edu.ksu.canvas.model.assignment.Submission> submission = submissionsList.stream().filter(sub -> sub.getUser().getId() == Integer.parseInt(canvasUserId)).findFirst();
+                if (assignmentExtended.isPresent() && submission.isPresent()) {
+                    AssignmentExtended assignment = assignmentExtended.get();
+                    int allowedAttempts = assignment.getAllowedAttempts();
+                    int attempt = 0;
+                    if (submission.get().getAttempt() != null && allowedAttempts >0) {
+                        attempt = submission.get().getAttempt();
+                        if (attempt >= allowedAttempts) {
```
Comment (machristie) [inline src/main/java/edu/iu/terracotta/service/app/impl/QuestionSubmissionServiceImpl.java:334]: So if I understand this, this gets the attempt number of the first submission for the user found in the list of submissions. But don't you want to get the maximum attempt number of all of the user's submissions?  Or alternatively just count the user's submissions?

That is, I assume that if a student has, say, 3 submissions, the API call with return 3 submissions objects for that student, and one of them will have attempt=1, one will have attempt=2 and one will have attempt=3. It's not clear to me that the first submission found for the user would be the third attempt with attempt=3, so seems like you need to take a maximum of them all.

File: src/main/java/edu/iu/terracotta/utils/TextConstants.java [hunk 3]
```diff
@@ -61,5 +61,6 @@ private TextConstants() {
     public static final String GROUP_PENDING = "Error 115: Group_Pending: The student has not been assigned to a group";
     public static final String ASSIGNMENT_LOCKED = "Error 116: The assignment is not open at this moment";
     public static final String LIMIT_OF_SUBMISSIONS_REACHED ="Error 117: You can't answer this assignment again";
+    public static final String MAX_SUBMISSION_ATTEMPTS_REACHED = "Error 118: Max submission attempts already reached";
```
Comment (machristie) [inline src/main/java/edu/iu/terracotta/utils/TextConstants.java:64]: Error 118 is already taken (https://github.com/terracotta-education/terracotta/blob/development/src/main/java/edu/iu/terracotta/controller/app/ConditionController.java#L179), can you use error code 150 instead?

(we have a separate issue to get these error codes documented, but see this comment for how to list out existing error codes: https://terracotta-education.atlassian.net/browse/TCOTA-358?focusedCommentId=10277)

**Review by machristie — CHANGES_REQUESTED** [review]
Thanks @isururanawaka . This is looking really good, although I do have some review comments.

Also, regarding the update to `updateSubmissions`, I saw you comment on Jira, and yes let's revert that and make the extra call from the frontend.

File: src/main/java/edu/iu/terracotta/service/app/impl/QuestionSubmissionServiceImpl.java (hunk 2 above)
Comment (isururanawaka) [reply src/main/java/edu/iu/terracotta/service/app/impl/QuestionSubmissionServiceImpl.java:334]: @machristie It seems it returns the highest attempt always.

**Review by machristie — APPROVED** [review]

**Pull request merged**